src/rgw/rgw_d3n_datacache.h

```diff
@@ -0,0 +1,366 @@
+  }
+
+  int flush_read_list(struct get_obj_data* d);
+  int get_obj_iterate_cb(const rgw_raw_obj& read_obj, off_t obj_ofs,
+                         off_t read_ofs, off_t len, bool is_head_obj,
+                         RGWObjState *astate, void *arg) override;
+};
+
+
+template<typename T>
+int D3nRGWDataCache<T>::flush_read_list(struct get_obj_data* d) {
+
+  d->d3n_datacache_lock.lock();
+  std::list<bufferlist> l;
+  l.swap(d->d3n_read_list);
+  d->d3n_read_list.clear();
+  d->d3n_datacache_lock.unlock();
+
+  int r = 0;
+
+  std::string oid;
+  std::list<bufferlist>::iterator iter;
+  for (iter = l.begin(); iter != l.end(); ++iter) {
+    bufferlist& bl = *iter;
+    oid = d->d3n_get_pending_oid();
+    if(oid.empty()) {
+      lsubdout(g_ceph_context, rgw, 0) << "ERROR: flush_read_list(): d3n_get_pending_oid() returned empty oid" << dendl;
+      r = -ENOENT;
+      break;
+    }
+    if (bl.length() <= 0x400000)
+      d3n_data_cache.put(bl, bl.length(), oid);
+  }
+
+  return r;
+}
+
+template<typename T>
+int D3nRGWDataCache<T>::get_obj_iterate_cb(const rgw_raw_obj& read_obj, off_t obj_ofs,
+                                 off_t read_ofs, off_t len, bool is_head_obj,
+                                 RGWObjState *astate, void *arg) {
+
+  librados::ObjectReadOperation op;
+  struct get_obj_data* d = static_cast<struct get_obj_data*>(arg);
+  string oid, key;
+
+  int r = 0;
+
+  if (is_head_obj) {
+    // only when reading from the head object do we need to do the atomic test
+    r = T::append_atomic_test(astate, op);
+    if (r < 0)
+      return r;
+
+    if (astate && obj_ofs < astate->data.length()) {
+      unsigned chunk_len = std::min((uint64_t)astate->data.length() - obj_ofs, (uint64_t)len);
+
+      r = d->client_cb->handle_data(astate->data, obj_ofs, chunk_len);
+      if (r < 0)
+        return r;
+
+      len -= chunk_len;
+      d->offset += chunk_len;
+      read_ofs += chunk_len;
+      obj_ofs += chunk_len;
+      if (!len)
+        return 0;
+    }
+  }
+
+  lsubdout(g_ceph_context, rgw, 20) << "D3nRGWDataCache::get_obj_iterate_cb oid=" << read_obj.oid << " obj-ofs=" << obj_ofs << " read_ofs=" << read_ofs << " len=" << len << dendl;
+  op.read(read_ofs, len, nullptr, nullptr);
+
+  const uint64_t cost = len;
+  const uint64_t id = obj_ofs; // use logical object offset for sorting replies
+  oid = read_obj.oid;
+
+  d->d3n_add_pending_oid(oid);
+
+  if (d3n_data_cache.get(read_obj.oid)) {
+    auto obj = d->store->svc.rados->obj(read_obj);
+    r = obj.open();
+    if (r < 0) {
+      lsubdout(g_ceph_context, rgw, 0) << "Error: d3n failed to open rados context for " << read_obj << ", r=" << dendl;
+      return r;
+    }
+    auto completed = d->aio->get(obj, rgw::Aio::cache_op(std::move(op), d->yield, obj_ofs, read_ofs, len, g_conf()->rgw_d3n_l1_datacache_persistent_path), cost, id);
+    if (g_conf()->rgw_d3n_l1_libaio_read) {
+      r = d->drain();
+    } else {
+      r = d->flush(std::move(completed));
+    }
```

Comment:
this needs to call `d->flush(std::move(completed));` unconditionally to write any buffers from prior async `get()` calls to the client

we should not call `drain()` here though, because it blocks until all outstanding requests finish. the point of the `AioThrottle` classes is to keep several IOs in flight. `drain()` is being called at the end of `RGWRados::Object::Read::iterate()`, after we've submitted all the necessary read ops

Reply:
quoting comment https://github.com/ceph/ceph/pull/36266#discussion_r550335978
> I have implemented new `cache_aio_abstract()` and `remote_aio_abstract()` for both Local (L1) and remote cache(L2-via s3) operations, which allow you to have `d->flush() `instead of `d->drain()` without any issue. You can find my changes in the commit down below.
> 
> [ekaynar/ceph-master@379ab0b](https://github.com/ekaynar/ceph-master/commit/379ab0bd45d925310f51b79c4e5e552992122eaf)
thank you very much @ekaynar

Reply:
changed from `drain()` to `flush()`